src/protocol/sort/shuffle.rs

```diff
@@ -31,6 +31,52 @@ pub struct Shuffle<'a, F, S> {
     step_fn: fn(ShuffleStep) -> S,
 }
 
+#[derive(Debug)]
+enum ShuffleOrUnshuffle {
+    Shuffle,
+    Unshuffle,
+}
+
+/// This implements Fisher Yates shuffle described here <https://en.wikipedia.org/wiki/Fisher%E2%80%93Yates_shuffle>
+#[allow(clippy::cast_possible_truncation, dead_code)]
+pub(self) fn generate_random_permutation(
+    batchsize: usize,
+    prss: &PrssSpace,
+) -> (Permutation, Permutation) {
+    // Chacha8Rng expects a [u8;32] seed whereas prss returns a u128 number.
+    // We are using two seeds from prss to generate a seed for shuffle and concatenating them
+    // Since reshare uses indexes 0..batchsize to generate random numbers from prss, we are using
+    // batchsize and batchsize+1 as index to get seeds for permutation
+    let randoms = (
+        prss.generate_values(batchsize as u128),
+        prss.generate_values(batchsize as u128 + 1),
+    );
+
+    // generate seed for shuffle
+    let (mut seed_left, mut seed_right) = (Vec::with_capacity(32), Vec::with_capacity(32));
+    seed_left.extend_from_slice(&randoms.0 .0.to_le_bytes());
+    seed_left.extend_from_slice(&randoms.1 .0.to_le_bytes());
+
+    seed_right.extend_from_slice(&randoms.0 .1.to_le_bytes());
+    seed_right.extend_from_slice(&randoms.1 .1.to_le_bytes());
+
+    let mut permutations: (Vec<usize>, Vec<usize>) =
+        ((0..batchsize).collect(), (0..batchsize).collect());
+    // shuffle 0..N based on seed
+    permutations
+        .0
+        .shuffle(&mut ChaCha8Rng::from_seed(seed_left.try_into().unwrap()));
+    permutations
+        .1
+        .shuffle(&mut ChaCha8Rng::from_seed(seed_right.try_into().unwrap()));
+
+    //take inverse and return
```

Comment:
Why do we need to take the inverse? If it's just a random shuffle, does it matter?

Reply:
doesnt matter, removing it